include/lug/System/Exception.hpp

```diff
@@ -0,0 +1,94 @@
+#include <cstdint>
+#include <string>
+#include <lug/System/Export.hpp>
+
+namespace lug {
+namespace System {
+
+class LUG_SYSTEM_API Exception: public std::exception {
+public:
+    Exception(const char *typeName, const std::string &description, const char* file, const char* function, uint32_t line);
+    Exception(Exception&& other) = default;
+    Exception(const Exception& other) = default;
+    Exception& operator=(const Exception& other) = default;
+    Exception& operator=(Exception&& other) = default;
+    ~Exception() throw() {}
+
+    const std::string& getTypeName() const;
+    const std::string& getDescription() const;
+    const std::string& getFile() const;
+    const std::string& getFunction() const;
+    const long getLine() const;
+
+    const char* what() const override;
+
+private:
+    std::string _typeName;
+    std::string _description;
+    std::string _file;
+    std::string _function;
+    uint32_t _line;
+
+    mutable std::string _fullDesc;
+};
+
+
+class LUG_SYSTEM_API NotImplementedException: public Exception {
+public:
+    NotImplementedException(const std::string &description, const char* file, const char* function, uint32_t line)
+    : Exception{"NotImplementedException", description, file, function, line} {}
+};
+
+
+class LUG_SYSTEM_API FileNotFoundException: public Exception {
+public:
+    FileNotFoundException(const std::string &description, const char* file, const char* function, uint32_t line)
+    : Exception{"FileNotFoundException", description, file, function, line} {}
+};
+
+
+class LUG_SYSTEM_API IOException: public Exception {
+public:
+    IOException(const std::string &description, const char* file, const char* function, uint32_t line)
+    : Exception{"IOException", description, file, function, line} {}
+};
+
+
+class LUG_SYSTEM_API InvalidParametersException: public Exception {
+public:
+    InvalidParametersException(const std::string &description, const char* file, const char* function, uint32_t line)
+    : Exception{"InvalidParametersException", description, file, function, line} {}
+};
+
+
+class LUG_SYSTEM_API InternalErrorException: public Exception {
+public:
+    InternalErrorException(const std::string &description, const char* file, const char* function, uint32_t line)
+    : Exception{"InternalErrorException", description, file, function, line} {}
+};
+
+
+class LUG_SYSTEM_API RendererException: public Exception {
+public:
+    RendererException(const std::string &description, const char* file, const char* function, uint32_t line)
+    : Exception{"RendererException", description, file, function, line} {}
+};
+
+
+class LUG_SYSTEM_API WindowException: public Exception {
+public:
+    WindowException(const std::string &description, const char* file, const char* function, uint32_t line)
+    : Exception{"WindowException", description, file, function, line} {}
+};
+
+
+#define LUG_EXCEPT(type, desc)\
+    do {\
+      throw type(desc, __FILE__, __FUNCTION__, __LINE__);\
```

Comment:
4 spaces * 2?
Maybe use `__PRETTY_FUNCTION__` (gcc & clang) and `__FUNCSIG__` (MSVC).